json_logic/__init__.py

```diff
@@ -1,201 +1,457 @@
-    not_found = object()
-    ret = []
-    for arg in args:
-        if get_var(data, arg, not_found) is not_found:
-            ret.append(arg)
-        else:
-            found += 1
-            if found >= min_required:
-                return []
-    return ret
+    return missing_array
 
 
-operations = {
-    "==": soft_equals,
-    "===": hard_equals,
-    "!=": lambda a, b: not soft_equals(a, b),
-    "!==": lambda a, b: not hard_equals(a, b),
-    ">": lambda a, b: less(b, a),
-    ">=": lambda a, b: less(b, a) or soft_equals(a, b),
-    "<": less,
-    "<=": less_or_equal,
-    "!": lambda a: not a,
-    "!!": bool,
-    "%": lambda a, b: a % b,
-    "and": lambda *args: reduce(lambda total, arg: total and arg, args, True),
-    "or": lambda *args: reduce(lambda total, arg: total or arg, args, False),
-    "?:": lambda a, b, c: b if a else c,
-    "if": if_,
-    "log": lambda a: logger.info(a) or a,
-    "in": lambda a, b: a in b if "__contains__" in dir(b) else False,
-    "cat": lambda *args: "".join(str(arg) for arg in args),
-    "+": plus,
-    "*": lambda *args: reduce(lambda total, arg: total * float(arg), args, 1),
-    "-": minus,
-    "/": lambda a, b=None: a if b is None else float(a) / float(b),
-    "min": lambda *args: min(args),
-    "max": lambda *args: max(args),
-    "merge": merge,
-    "count": lambda *args: sum(1 if a else 0 for a in args),
+data_operations = {
+    'var': _var,
+    'missing': _missing,
+    'missing_some': _missing_some
 }
 
 
-def jsonLogic(tests, data=None):
-    """Executes the json-logic with given data."""
-    # You've recursed to a primitive, stop!
-    if tests is None or not isinstance(tests, dict):
-        return tests
+# MAIN LOGIC
 
-    data = data or {}
+def jsonLogic(logic, data=None):
+    """
+    Evaluate provided JsonLogic using given data (if any).
+    If a single JsonLogic entry is provided - return a single resulting value.
+    If an array of JsonLogic entries is provided - return an array of each
+    entry's resulting values.
+    """
 
-    operator = list(tests.keys())[0]
-    values = tests[operator]
+    # Is this an array of JsonLogic entries?
+    if isinstance(logic, (list, tuple)):
+        return list(map(lambda subset: jsonLogic(subset, data), logic))
 
-    # Easy syntax for unary operators, like {"var": "x"}
+    # You've recursed to a primitive, stop!
+    if logic is None or not isinstance(logic, dict):
+        return logic
+
+    # Get operator
+    operators = list(logic.keys())
+    if len(operators) != 1:
+        raise ValueError("Each rule must only contain a single operator.")
+    operator = operators[0]
+
+    # Get values
+    values = logic[operator]
+    # Easy syntax for unary operators like {"var": "x"}
     # instead of strict {"var": ["x"]}
-    if not isinstance(values, list) and not isinstance(values, tuple):
+    if not isinstance(values, (list, tuple)):
         values = [values]
 
+    # Get data
+    data = data or {}
```

Comment:
This should be set as the default in the function definition.
As in:
 def jsonLogic(logic, data={}):
It is cleaner and accomplishes the same thing.